cpp/src/arrow/compute/kernels/sum-internal.h

```diff
@@ -0,0 +1,204 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/aggregate.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+
+namespace arrow {
+
+class Array;
+class DataType;
+
+namespace compute {
+
+// Find the largest compatible primitive type for a primitive type.
+template <typename I, typename Enable = void>
+struct FindAccumulatorType {};
+
+template <typename I>
+struct FindAccumulatorType<I, enable_if_signed_integer<I>> {
+  using Type = Int64Type;
+};
+
+template <typename I>
+struct FindAccumulatorType<I, enable_if_unsigned_integer<I>> {
+  using Type = UInt64Type;
+};
+
+template <typename I>
+struct FindAccumulatorType<I, enable_if_floating_point<I>> {
+  using Type = DoubleType;
+};
+
+template <typename ArrowType, typename StateType>
+class SumAggregateFunction final : public AggregateFunctionStaticState<StateType> {
+  using CType = typename TypeTraits<ArrowType>::CType;
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+  static constexpr int64_t kTinyThreshold = 32;
+  static_assert(kTinyThreshold > 18,
+                "ConsumeSparse requires at least 18 elements to fit 3 bytes");
+
+ public:
+  Status Consume(const Array& input, StateType* state) const override {
+    const ArrayType& array = static_cast<const ArrayType&>(input);
+
+    if (input.null_count() == 0) {
+      *state = ConsumeDense(array);
+    } else if (input.length() <= kTinyThreshold) {
+      // In order to simplify ConsumeSparse implementation (requires at least 3
+      // bytes of bitmap data), small arrays are handled differently.
+      *state = ConsumeTiny(array);
+    } else {
+      *state = ConsumeSparse(array);
+    }
+
+    return Status::OK();
+  }
+
+  Status Merge(const StateType& src, StateType* dst) const override {
+    *dst += src;
+    return Status::OK();
+  }
+
+  Status Finalize(const StateType& src, Datum* output) const override {
+    *output = src.Finalize();
+    return Status::OK();
+  }
+
+  std::shared_ptr<DataType> out_type() const override { return StateType::out_type(); }
+
+ private:
+  StateType ConsumeDense(const ArrayType& array) const {
```

Comment:
This (and `ConsumeSparse`, `ConsumeTiny`) looks pretty handy, maybe later we could refactor it into `BitUtil::VisitBitmap()` or so

Reply:
+1

Reply:
Soon the complexity of future needs will outgrow the current implementation, e.g. we'll need to mix with selection vectors/bitmap, groups, etc. I intend to open a discussion in the mailing list about providing iterator facilities with many options (batch, masked, groups).